Source/cmLocalGenerator.cxx

```diff
@@ -879,6 +879,27 @@ void cmLocalGenerator::AddCompileOptions(std::string& flags,
     }
   }
   this->AddCompilerRequirementFlag(flags, target, lang);
+
+  cmMakefile* mf = this->GetMakefile();
+  if ((mf->GetDefinition("MSVC_C_ARCHITECTURE_ID") ||
+      mf->GetDefinition("MSVC_CXX_ARCHITECTURE_ID") ||
+      mf->GetDefinition("MSVC_CUDA_ARCHITECTURE_ID"))) {
+    // If the target is a Managed C++ one, /ZI is not compatible.
+
+    if (flags.find("/clr") != std::string::npos ||
+        flags.find("-clr") != std::string::npos ||
+          target->GetManagedType(config) ==
+            cmGeneratorTarget::ManagedType::Mixed ||
+        target->GetManagedType(config) ==
+          cmGeneratorTarget::ManagedType::Managed)
+      {
+      // Since /ZI is the default for DEBUG configuration, replace it with /Zi
+      // when the project is a managed one (i.e. /clr passed in).
+      if (cmSystemTools::UpperCase(config) == "DEBUG") {
+        cmSystemTools::ReplaceString(flags, " /ZI ", " /Zi ");
```

Comment:
Can we safely assume that the trailing space will always be there?

Reply:
yes, that is a flaw of this string replacement, it wont work if /ZI is defined at beginning or end of the string containing the flags. Nevertheless by default /ZI defined to be with a space before and after in CMAKE_CXX_FLAGS_DEBUG_INIT, and our intent is to replace the flag as being defined in the default configuration. Failing at removing the /ZI flag when the user project manipulated the flags should be acceptable, as the manipulation should have already doing the right substitution already.